### ISSUE TITLE
Indent Performance: string/comment lookups

### ISSUE THREAD

**Issue by jbodah** [elixir-editors/vim-elixir#407]
Given the following file:

```
defmodule Http do
  defmacro __using__(_) do
    quote do
      use HTTPoison.Base

      def get(url, headers \\ [], options \\ []) do
        response(:get, url, "", headers, options)
      end

      def delete(url, headers \\ [], options \\ []) do
        response(:delete, url, "", headers, options)
      end

      def post(url, body, headers \\ [], options \\ []) do
        response(:post, url, body, headers, options)
      end

      def patch(url, body, headers \\ [], options \\ []) do
        response(:patch, url, body, headers, options)
      end

      def response(method, url, body, headers \\ [], options \\ []) do
        url = options.base_url <> url
        log(method, :start, url, options)
        start_time = now()

        response = request(method, url, body, headers, options)

        case response do
          {:ok, response} ->
            log(method, :finish, url, options, response, start_time)

          {:error, response} ->
            log(method, :error, url, options, response, start_time)
        end

        response
      end

      def process_request_options(%_{} = options) do
        options
        |> Map.from_struct()
        |> Enum.into([])
        |> process_request_options()
      end

      def process_request_options(options) do
        [
          timeout: 10_000,
          recv_timeout: 15_000,
          hackney: [pool: :http_default_pool]
        ]
        |> Keyword.merge(options)
      end

      def process_request_body(body) when is_map(body), do: Poison.encode!(body)
      def process_request_body(body) when is_binary(body), do: body

      def process_response_body(body) do
        case Poison.Parser.parse(body) do
          {:ok, parsed_body} -> parsed_body
          {:error, _} -> body
        end
      end

      def process_request_headers(headers) do
        headers
      end

      defp now(), do: DateTime.utc_now() |> DateTime.to_unix(:millisecond)

      defp log(request_method, at, path, options, response \\ nil, started_at \\ nil) do
        data = %{
          api_request: true,
          at: at,
          host: options.base_url,
          request_id: options.request_id,
          url: path,
          method: request_method,
          path: options.generic_path,
          elapsed: started_at && now() - started_at
          # status: response && response.status_code,
          # reason: response && response.reason
        }

        # remove keys with falsy values
        data = for {key, value} <- data, value, into: %{}, do: {key, value}

        case response do
          nil -> Plexy.Logger.debug(data)
          %HTTPoison.Response{} -> Plexy.Logger.debug(data)
          %HTTPoison.Error{} = error -> Plexy.Logger.error(data)
        end
      end
    end
  end
end
```

We get the following profiler output:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
 9466             10.457848  <SNR>77_is_string_or_comment()
  312  11.036247   1.915420  elixir#indent#handle_inside_nested_construct()
 8828   9.363995   0.120360  elixir#indent#searchpair_back_skip()
 1522   1.123686   0.106535  <SNR>77_ends_with()
  384  13.163921   0.071142  elixir#indent#indent()
   52   0.205268   0.054884  elixir#indent#handle_inside_generic_block()
  382   0.210430   0.053353  elixir#indent#handle_starts_with_end()
 2542   0.096197   0.048705  <SNR>77_starts_with()
  398  61.177775   0.024607  <SNR>77_get_base_indent()
  326   0.164678   0.021031  elixir#indent#handle_starts_with_close_bracket()
 4644              0.017567  <SNR>77_debug()
  312   0.018112   0.006850  elixir#indent#handle_starts_with_binary_operator()
  334   0.020190   0.006785  elixir#indent#handle_starts_with_pipe()
  358   0.013070   0.005970  elixir#indent#handle_starts_with_mid_or_end_block_keyword()
  358   0.087979   0.005809  elixir#indent#handle_following_trailing_do()
  334   0.037895   0.003777  elixir#indent#handle_following_trailing_binary_operator()
  230              0.003508  elixir#indent#handle_starts_with_comment()
    8              0.003288  <SNR>77_find_last_pos()
  384  13.166979   0.003058  elixir#indent()
 1274              0.002925  <SNR>77_keyword()
```

The clear culprit is delegating to the syntax engine over and over. One thought - can we filter out all of the strings in a preprocessing step perhaps? That would surely be cheaper than checking them over and over

**Comment by kswope** [comment]
This one is brutal for me.  I don't have fancy macbook pros like you guys, I've got a $200 chromebook.  This problem slows down almost everything when loading a new file, even the cursor movement!  When loading a new file (small file) I have to wait about 10 seconds for all lag to stop.

**Comment by jbodah** [comment]
@kswope sorry to hear that. This has been a real thorn in my side as I don't really have a good solution that isn't going to take a long time to implement without introducing regressions.

A few questions:

* Are you using `vim-elixir` directly (I've seen some old copies pulled in through other libraries)? If so, are you on HEAD?

* Which version of vim/nvim are you using?

* Which `regexpengine` are you using (`:h regexpengine`)?

* Could you send me a profile of one of your sessions?

**Comment by kswope** [comment]
> Are you using vim-elixir directly (I've seen some old copies pulled in through other libraries)? If so, are you on HEAD?

```
Plug 'elixir-lang/vim-elixir' 
```

> Which version of vim/nvim are you using?
```
(xenial)kevin@localhost:~/wheatt-phx$ vim --version
NVIM v0.2.2
Build type: RelWithDebInfo
LuaJIT 2.0.4
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=2 -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim-HJnLxJ/neovim-0.2.2/build/config -I/build/neovim-HJnLxJ/neovim-0.2.2/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim-HJnLxJ/neovim-0.2.2/build/src/nvim/auto -I/build/neovim-HJnLxJ/neovim-0.2.2/build/include
Compiled by pkg-vim-maintainers@lists.alioth.debian.org

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```


> Which regexpengine are you using (:h regexpengine)?
Does this tell us anything?
```
   'regexpengine' 're'     number  (default 0)                                                                   
                           global                                                                                
           This selects the default regexp engine. two-engines                                                   
           The possible values are:                                                                              
                   0       automatic selection                                                                   
                   1       old engine                                                                            
                   2       NFA engine  
```

> Could you send me a profile of one of your sessions?

All I did was open vim, search for an elixir file, and open it.

[profile.log](https://github.com/elixir-editors/vim-elixir/files/1826816/profile.log)

**Comment by jbodah** [comment]
> All I did was open vim, search for an elixir file, and open it.
>
> profile.log

Are you experiencing lag here? None of these jump out as vim-elixir to me (although I could be mistaken). How is it without vim-airline?

```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   0.721029   0.101978  <SNR>78_BuildPrompt()
    8   0.653305   0.000420  <SNR>78_PrtFocusMap()
    8   0.652885   0.000694  <SNR>78_PrtAdd()
    4   0.594375   0.003286  <SNR>70_LocalBrowseRefresh()
   50   0.592981   0.023987  airline#check_mode()
    4   0.564470   0.049523  airline#highlighter#highlight()
  349   0.544743   0.145172  airline#highlighter#exec()
  749   0.504361   0.197983  airline#highlighter#get_highlight()
    4   0.416102   0.001357  <SNR>70_NetrwRefresh()
    4   0.414746   0.001171  netrw#LocalBrowseCheck()
    4   0.412700   0.005920  <SNR>70_NetrwBrowse()
    4   0.371339   0.008598  <SNR>70_PerformListing()
  156   0.358482   0.028945  <SNR>62_exec_separator()
   11   0.351361   0.031680  9()
    9   0.317156   0.003022  <SNR>78_Update()
    9   0.298603   0.002905  ctrlp#statusline()
    9   0.295406   0.002234  airline#extensions#ctrlp#ctrlp_airline()
    1   0.280450   0.001379  <SNR>78_AcceptSelection()
 1498   0.278046             <SNR>62_get_syn()
    1   0.277177   0.001165  ctrlp#acceptfile()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
 1498              0.278046  <SNR>62_get_sun()
  749   0.504361   0.197983  airline#highlighter#get_highlight()
  349   0.544743   0.145172  airline#highlighter#exec()
    4   0.133336   0.132884  ctrlp#utils#writecache()
    9   0.721029   0.101978  <SNR>78_BuildPrompt()
  534   0.154767   0.081180  <SNR>78_matchlens()
    1   0.069861   0.069813  <SNR>78_UserCmd()
  349              0.061462  <SNR>62_CheckDefined()
    3   0.064881   0.050648  ale#sign#SetSigns()
    4   0.564470   0.049523  airline#highlighter#highlight()
   13   0.052180   0.048620  <SNR>10_LoadFTPlugin()
  322              0.044812  <SNR>34_set()
  500              0.039605  <SNR>34_get()
  122   0.230540   0.039347  <SNR>78_mixedsort()
   13   0.043604   0.037588  <SNR>33_Unmap()
    4              0.033514  <SNR>70_NetrwSetSort()
   11   0.351361   0.031680  9()
    4   0.053555   0.031627  <SNR>70_LocalListing()
  322   0.107255   0.030402  <SNR>33_option_init()
    6   0.030453   0.030303  ale#job#Start()
```

**Comment by kswope** [comment]
Here is something

Fast!
```
" Plug 'w0rp/ale'                                                                                           
" Plug 'elixir-lang/vim-elixir'
```

Fast!
```
Plug 'w0rp/ale'                                                                                             
" Plug 'elixir-lang/vim-elixir'
```

Fast!
```
" Plug 'w0rp/ale'                                                                                           
Plug 'elixir-lang/vim-elixir'
```

Painfully slow!
```
Plug 'w0rp/ale'                                                                                             
Plug 'elixir-lang/vim-elixir'
```

Note that I haven't had any problems with ale and other language plugins.  I can even switch to a js file in my elixir project and the lag doesn't happen.

Another clue, and I wish I could make a video of this.  When ale is ready to do its business, its shifts the code over to the right a few spaces to make room for the markers.  This usually happens right away with other filetypes.  But when I load an elixir file, the lag happens for about 10 seconds, and ends when the shift happens.  Its like ale has an fight with vim-elixir, the fight eventually ends, the code shifts right, and the lag is gone.

**Comment by jbodah** [comment]
Nice find - I'll try to dig into this a bit and see if I can find anything. Nothing rings a bell looking at the code (I don't use ale myself, but I noticed @hauleth has contributed there and has also contributed to vim-elixir - not sure if he's also used them in conjunction and seen weird behavior)

**Comment by hauleth** [comment]
As an author of Ale's Credo linter my suggestion would be: don't. I can (and probably will) look for the source of the issue, however I had bad memories from time when I used Elixir with Ale. The main problem is that you need to compile whole project before running linter like Credo, Dogma or Dialyzer. All due to fact how Mix works.

About your test matrix. It isn't surprise that these two:

> Fast!
>
> ```
> Plug 'w0rp/ale'
> " Plug 'elixir-lang/vim-elixir'
> ```
>
> Fast!
>
> ```
> " Plug 'w0rp/ale'
> Plug 'elixir-lang/vim-elixir'
> ```

Are fast. In first one you will not run Ale linter, due to fact that Vim doesn't know about Elixir filetype, so by the definition, neither do Ale. So as the filetype is unknown, no linter is ran. In the second example there is no Ale, so nothing to be slow at all.

Only solution I know of that could help there is rrrene/credo#473 or use ElixirLS diagnostics.

**TL;DR**

This is due project rebuild on each run.

**Comment by Coffei** [comment]
I also believe I experience this issue. On as simple file as this https://gist.github.com/Coffei/1ad18c03a2624e37144e5c0ba622cdd8 I can make my vim come to a complete halt just by holding any letter in insert mode for under a second. Navigation is fast, it only affects typing.

I've tried running clean vim with only `elixir-vim` but cannot reproduce it to that extent, so it may be some combination with other plugins. What I observed though, is the issue is triggered by `set foldmethod=syntax`. When I don't do that it runs fine.

I'm not sure what could help you so let me know what info you need.

**Comment by jbodah** [comment]
> What I observed though, is the issue is triggered by set foldmethod=syntax

This aligns with other measures that the syntax rules are computationally expensive

**Comment by jbodah** [comment]
Linking as potentially relevant: https://github.com/neovim/neovim/issues/8607#issuecomment-399348632

**Comment by Qizot** [comment]
> I can make my vim come to a complete halt just by holding any letter in insert mode for under a second. Navigation is fast, it only affects typing.

I've just encountered the same problem. I have no idea what other plugin may interfere with this and there are to many to investigate that.

**Comment by jbodah** [comment]
@Qizot could you provide me an example file and repro scenario? Also could you tell me which `vim --version` you're using? I just tried reproing the issue here but could not https://github.com/elixir-editors/vim-elixir/issues/407#issuecomment-400543487

**Comment by Qizot** [comment]
> @Qizot could you provide me an example file and repro scenario? Also could you tell me which `vim --version` you're using? I just tried reproing the issue here but could not [#407 (comment)](https://github.com/elixir-editors/vim-elixir/issues/407#issuecomment-400543487)

I don't know if that matters (might a lot) but I'm using neovim versioned `NVIM v0.5.0-46e7414`  with iterm on macos.
To reproduce that you can use mentioned file and go to insert mode anywhere and long press enter key.  Sometimes it works well and sometimes there is a huge lag and after that dozens of new lines appear (as it would by simply holding the key), but after that you can press enter normally and it works (as far as you don't change cursor position). The enter key is quite noticeable and there is a visible little lag when using 'O' to start in insert mode on the new line, can't tell you much more as I've been using vs code lately.
